agent/sidecar_service.go

```diff
@@ -0,0 +1,170 @@
+// registration. This will be the same as the token parameter passed unless the
+// SidecarService definition contains a distint one.
+func (a *Agent) sidecarServiceFromNodeService(ns *structs.NodeService, token string) (*structs.NodeService, []*structs.CheckType, string, error) {
+	if ns.Connect.SidecarService == nil {
+		return nil, nil, "", nil
+	}
+
+	// Start with normal conversion from service definition
+	sidecar := ns.Connect.SidecarService.NodeService()
+
+	// Override the ID which must always be consistent for a given outer service
+	// ID. We rely on this for lifecycle management of the nested definition.
+	sidecar.ID = a.sidecarServiceID(ns.ID)
+
+	// Set some meta we can use to disambiguate between service instances we added
+	// later and are responsible for deregistering.
+	if sidecar.Meta != nil {
+		// Meta is non-nil validate it before we add the special key so we can
+		// enforce that user cannot add a consul- prefix one.
+		if err := structs.ValidateMetadata(sidecar.Meta, false); err != nil {
+			return nil, nil, "", err
+		}
+	}
+
+	// Flag this as a sidecar - this is not persisted in catalog but only needed
+	// in local agent state to disambiguate lineage when deregistereing the parent
+	// service later.
+	sidecar.LocallyRegisteredAsSidecar = true
+
+	// See if there is a more specific token for the sidecar registration
+	if ns.Connect.SidecarService.Token != "" {
+		token = ns.Connect.SidecarService.Token
+	}
+
+	// Setup some sane connect proxy defaults.
+	if sidecar.Kind == "" {
+		sidecar.Kind = structs.ServiceKindConnectProxy
+	}
+	if sidecar.Service == "" {
+		sidecar.Service = ns.Service + "-sidecar-proxy"
+	}
+	if sidecar.Address == "" {
+		// Inherit address from the service if it's provided
+		sidecar.Address = ns.Address
+	}
+	// Proxy defaults
+	if sidecar.Proxy.DestinationServiceName == "" {
+		sidecar.Proxy.DestinationServiceName = ns.Service
+	}
+	if sidecar.Proxy.DestinationServiceID == "" {
+		sidecar.Proxy.DestinationServiceID = ns.ID
+	}
+	if sidecar.Proxy.LocalServiceAddress == "" {
+		sidecar.Proxy.LocalServiceAddress = "127.0.0.1"
+	}
+	if sidecar.Proxy.LocalServicePort < 1 {
+		sidecar.Proxy.LocalServicePort = ns.Port
+	}
+
+	// Allocate port if needed (min and max inclusive).
+	rangeLen := a.config.ConnectSidecarMaxPort - a.config.ConnectSidecarMinPort + 1
+	if sidecar.Port < 1 && a.config.ConnectSidecarMinPort > 0 && rangeLen > 0 {
+		// This did pick at random which was simpler but consul reload would assign
+		// new ports to all the sidecar since it unloads all state and re-populates.
+		// It also made this more difficult to test (have to pin the range to one
+		// etc.). Instead we assign sequentially, but rather than N^2 lookups, just
+		// iterated services once and find the set of used ports in allocation
+		// range. We could maintain this state permenantly in agent but it doesn't
+		// seem to be necessary - even with thousands of services this is not
+		// expensive to compute.
```

Comment:
Possibly if you stored it locally then you could insulate yourself against the effects of restarting the agent with new fascinating config settings that would make this iteration perhaps re-port a sidecar that's already running.

Reply:
True, I'll leave that as future work if it turns out to be a real problem though. adding state adds more to maintain etc.